test(docs-infra): add missing unit tests for backoff example

sonukapoor opened the pull request:
This commit adds the missing unit test for the backoff example.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Documentation content changes

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

sonukapoor commented on aio/content/examples/practical-observable-usage/src/backoff.ts:
```diff
@@ -1,23 +1,33 @@
 }
 
+// #enddocregion
+function makeAjaxCall() {
```
@gkalpak I had to introduce the `#docplaster` and `#docregion` to this file to ensure that the `ajax` method at line 22 does not get called when running the unit tests.

gkalpak reviewed (changes requested):
Excellent! Testing ftw :tada:

Can you please split it up into two commits, one adding the test (with type `test`) and one fixing the source code (with type `fix`).

sonukapoor committed:
fix(docs-infra): fix the backoff example

Previously, the `backoff()` example did not work as intended. More
specifically, the `range(1, maxTries)` observable would complete
immediately after emitting the `maxTries`th value, causing the overall
observable to also complete. As a result, it would only make
`maxTries - 1` attempts to recover from an error. More importantly, the
outer observable would complete successfully instead of erroring.

This commit fixes the `backoff()` operator by ensuring it makes exactly
`maxTries` attempts to recover and it propagates the actual error to the
outer observable.

The test for this change is added in the next commit.

sonukapoor commented on aio/content/examples/practical-observable-usage/src/backoff.spec.ts:
```diff
@@ -0,0 +1,67 @@
+    jasmine.clock().tick(10);
+    expect(mockConsole.log.calls.allArgs()).toEqual([['Emitted: 0']]);
+
+    jasmine.clock().tick(10);
```
I am not sure what kind of assertions we should add. Can you advise?

gkalpak replied:
```suggestion
    jasmine.clock().tick(10);
    expect(mockConsole.log.calls.allArgs()).toEqual([['Emitted: 0']]);
```

sonukapoor commented on aio/content/examples/practical-observable-usage/src/backoff.spec.ts:
```diff
@@ -0,0 +1,67 @@
+
+    // First re-attempt after 250ms:
+    // Errors again on second emission and schedules retrying (with larger delay).
+    jasmine.clock().tick(100);
```
Oops. This should be actually `250` as it is the first attempt right?

gkalpak replied:
No this is actually correct. I realized I was looking at the wrong line, but then forgot to remove the comment.
It's the comment above that is wrong. It should say `after 100ms`, since we are using an 100ms delay for `backoff()`.

sonukapoor committed:
test(docs-infra): add missing unit tests for backoff example

This commit adds the missing unit test for the backoff example.

sonukapoor commented:
@gkalpak I have removed the `handleData` method and add the necessary block into the subscribe.

gkalpak reviewed (changes requested):
Thx for making the changes, @sonukapoor! I've left a couple of final minor comments.

Regarding the commits:
- I think it would be better to change their order. This way, it is easy to see that the fixed `backoff()` code works as expected. (Alternatively, you can mention in the `fix` commit that a test will be added in the next commit.)
- In the `fix` commit, can you please mention the following (to give more context on what exactly was broken):
    > Previously, the \`backoff()\` example did not work as intended. More
    > specifically, the \`range(1, maxTries)\` observable would complete
    > immediately after emitting the \`maxTries\`th value, causing the overall
    > observable to also complete. As a result, it would only make
    > \`maxTries - 1\` attempts to recover from an error. More importantly, the
    > outer observable would complete successfully instead of erroring.
    >
    > This commit fixes the \`backoff()\` operator by ensuring it makes exactly
    > \`maxTries\` attempts to recover and it propagates the actual error to the
    > outer observable.

sonukapoor commented on aio/content/examples/practical-observable-usage/src/backoff.ts:
```diff
@@ -1,23 +1,39 @@
+/**
+ * This function declaration is necessary to ensure that it does not get called
+ * when running the unit tests. It will not get rendered into the docs.
+ * The indention needs to start in the most level position as well because of how
```
Yes. I have added that.

sonukapoor commented on aio/content/examples/practical-observable-usage/src/backoff.ts:
```diff
@@ -1,23 +1,34 @@
-
-function handleData(data) {
-  // ...
+  .subscribe(data => () => {
```
Oops. Not sure what I was thinking. I have corrected this.

sonukapoor commented:
> * I think it would be better to change their order. This way, it is easy to see that the fixed `backoff()` code works as expected. (Alternatively, you can mention in the `fix` commit that a test will be added in the next commit.)

If we switch the order, then the test will break, because the `backoff.ts` will not include the `docRegionAjaxCall` function and it will try to make an `ajax` call. I have added a comment into the `fix` as you suggested. Thank you also for providing the additional information to add to the fix commit.

gkalpak commented on aio/content/examples/practical-observable-usage/src/backoff.ts:
```diff
@@ -1,23 +1,32 @@
 }
 
+// #enddocregion
+/**
```
I don't think we need to make this a jsdoc comment:

```suggestion
/*
```

gkalpak commented:
> If we switch the order, then the test will break, because the backoff.ts will not include the docRegionAjaxCall function and it will try to make an ajax call.

Not sure what you mean. (I am aware that the test will be failing, but not because of the reasons you mention - unless I am missing something 😕)

That's OK imo given that the code was broken before (it's not that the commit will cause the code to break - it will just highlight the broken logic) and we will be fixing it in the next commit.

But either order is fine (as long as there is a comment in the commit message).

gkalpak reviewed (changes requested):
The change lgtm, but they are not split correctly across the two commits. Please make sure that the fix commit only contains the relevant changes.

gkalpak reviewed (changes requested):
@sonukapoor, you have to break up the changes so that one commit contains only the changes that are related to the fix and the other contains all the changes (in both files) that are related to adding tests.

gkalpak reviewed (changes requested):
Now the code is split up correctly, but you seem to have undone some of the previous changes. Here is the diff of the commit before my last review and the current state of the PR: [143dea3..b294594][1]

Can you please re-revert these changes (while still keeping the code correctly split up)? 🙏  

[1]: https://github.com/angular/angular/compare/143dea3a241d1dae8295d49362f700d27814c333..b2945945107ea5e2d07f14feb99f0ee58d43bd65

sonukapoor committed:
fixup! test(docs-infra): add missing unit tests for backoff example

gkalpak reviewed (approved):
LGTM 🚀

gkalpak reviewed (approved):
Reviewed-for: global-docs-approvers
